File: day7/part2.py
# Read file input
with open("input.txt", "r") as f:
    data = [int(i) for i in f.read().rstrip('\n').split(',')]

largest = max(data)

# Total cost of fuel for each possible position
fuel = [0]*(largest+1)

# Calculate fuel cost for each position, one crab at a time
for crab_pos in data:
    # Fill right
    cost = 1
    for i in range(crab_pos+1,len(fuel),1):
        fuel[i] += (cost*(cost+1))//2
        cost+=1
    # Fill left
    cost = 1
    for i in range(crab_pos-1,-1,-1):
        fuel[i] += (cost*(cost+1))//2
        cost+=1

ans = min(fuel)

print(f"Result: {ans}")

File: day7/part2.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

int main() {
	freopen("input.txt","r",stdin);

	// Process crab data
	vector<int> data;
	while (cin) {
		int fish;
		cin >> fish;
		data.push_back(fish);
		// Ignore next character (comma)
		getchar();
	}
	data.pop_back();

	int largest = *max_element(data.begin(),data.end());

	// Total cost of fuel for each possible position
	vector<int> fuel(largest+1);

	// Calculate fuel cost for each position, one crab at a time
	for (int crab_pos : data) {
		int cost;
		// Fill right
		cost = 1;
		for (int i=crab_pos+1;i<largest+1;i++) {
			fuel[i] += (cost*(cost+1))/2;
			++cost;
		}
		// Fill left
		cost = 1;
		for (int i=crab_pos-1;i>=0;i--) {
			fuel[i] += (cost*(cost+1))/2;
			++cost;
		}
	}

	int ans = *min_element(fuel.begin(),fuel.end());

	cout << "Result: " << ans << '\n';
}
